aot/src/main/java/com/dylibso/chicory/aot/AotAnalyzer.java

```diff
@@ -0,0 +1,731 @@
+                break;
+            case GLOBAL_SET:
+            case LOCAL_SET:
+                // [t] -> []
+                stack.pop(stack.peek());
+                break;
+            case LOCAL_GET:
+                // [] -> [t]
+                stack.push(localType(functionType, body, (int) ins.operand(0)));
+                break;
+            case GLOBAL_GET:
+                // [] -> [t]
+                stack.push(globalTypes.get((int) ins.operand(0)));
+                break;
+            case DATA_DROP:
+            case ELEM_DROP:
+                // [] -> []
+                break;
+            default:
+                throw new ChicoryException("Unhandled opcode: " + ins.opcode());
+        }
+        out.add(new AotInstruction(AotOpCode.of(ins.opcode()), ins.operands()));
+    }
+
+    private static void updateStack(TypeStack stack, FunctionType functionType) {
+        for (ValueType type : reversed(functionType.params())) {
+            stack.pop(type);
+        }
+        for (ValueType type : functionType.returns()) {
+            stack.push(type);
+        }
+    }
+
+    private Optional<AotInstruction> unwindStack(
+            FunctionType functionType,
+            FunctionBody body,
+            AnnotatedInstruction ins,
+            int label,
+            TypeStack stack) {
+
+        boolean forward = true;
+        var target = body.instructions().get(label);
+        if (target.address() <= ins.address()) {
+            // the loop block is the instruction before the target
+            target = body.instructions().get(label - 1);
+            forward = false;
+        }
+        var scope = target.scope();
+
+        FunctionType blockType;
+        if (scope.opcode() == OpCode.END) {
+            // special scope for the function's implicit block
+            scope = FUNCTION_SCOPE;
+            blockType = functionType;
+        } else {
+            blockType = blockType(scope);
+        }
+
+        var types = forward ? blockType.returns() : blockType.params();
+        int keep = types.size();
+
+        // for a backward jump, the initial loop parameters are dropped
+        int drop = stack.types().size() - stack.scopeStackSize(scope);
+
+        // do not drop the return values for a forward jump
+        if (forward) {
+            drop -= types.size();
+        }
+
+        if (drop <= 0) {
+            return Optional.empty();
+        }
+
+        // operands: [drop, drop_types..., keep_types...]
+        var operands = LongStream.builder();
+        operands.add(drop);
+
+        List<ValueType> dropKeepTypes =
+                stack.types().stream().limit(drop + keep).collect(toCollection(ArrayList::new));
+        reverse(dropKeepTypes);
+        dropKeepTypes.stream().mapToLong(ValueType::id).forEach(operands::add);
+
+        return Optional.of(new AotInstruction(AotOpCode.DROP_KEEP, operands.build().toArray()));
+    }
+
+    private FunctionType blockType(Instruction ins) {
```

Comment:
`blockType` is used also for `LOOP`s which is using the params instead of the returns, am I missing something?

Reply:
This is how block types are encoded per the [specification](https://webassembly.github.io/spec/core/syntax/instructions.html#syntax-instr-control):

> A structured instruction can consume input and produce output on the operand stack according to its annotated block type. It is given either as a *type index* that refers to a suitable *function type*, or as an optional *value type* inline, which is a shorthand for the function type `[] -> [valuetype?]`.